### PR TITLE
API for object passing between platform threads.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+fun <T> detachObjectGraph(mode: TransferMode = TransferMode.CHECKED, producer: () -> T): COpaquePointer? =
+        @Suppress("UNCHECKED_CAST")(detachObjectGraphInternal(mode.value, producer as () -> Any?))
+
+/**
+ * Atatched previously detached object subgraph.
+ * Please note, that once object graph is attached, the stable pointer does not have sense anymore,
+ * and shall be discarded.
+ */
+fun <T> attachObjectGraph(stable: COpaquePointer?): T =
+        @Suppress("UNCHECKED_CAST")(attachObjectGraphInternal(stable) as T)
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: It makes some sense to make this cast checked, by making `T` reified.

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * Creates stable pointer to object, ensuring associated object subgraph is disjoint in checked mode.
+ * It could be stored to C variable or passed to another thread, where it could be retrieved with attachObjectGraph().
+ */
+fun <T> detachObjectGraph(mode: TransferMode = TransferMode.CHECKED, producer: () -> T): COpaquePointer? =
+        @Suppress("UNCHECKED_CAST")(detachObjectGraphInternal(mode.value, producer as () -> Any?))
+
+/**
+ * Atatched previously detached object subgraph.
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: Atatched  -> Attache**s**

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+fun <T> T.shallowCopy(): T = @Suppress("UNCHECKED_CAST") (shallowCopyInternal(this) as T)
+
+/**
+ * Creates verbatim *deep* copy of passed object's graph, use *VERY* carefully to create disjoint object graph.
+ * Note that this function could potentially duplicate a lot of objects.
+ */
+fun <T> T.deepCopy(): T = TODO()
+
+/**
+ * Creates stable pointer to object, ensuring associated object subgraph is disjoint in checked mode.
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: in checked mode -> in specified mode ([TransferMode.CHECKED] by default).

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * It could be stored to C variable or passed to another thread, where it could be retrieved with attachObjectGraph().
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: `attachObjectGraph()` -> `[attachObjectGraph]`

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt (hunk 1 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: Done

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt (hunk 2 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: Done

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt (hunk 3 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: Done.

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt (hunk 4 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: Done

File: runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt [hunk 5]
```diff
@@ -58,19 +58,20 @@ fun <T> T.shallowCopy(): T = @Suppress("UNCHECKED_CAST") (shallowCopyInternal(th
...
 /**
- * Atatched previously detached object subgraph.
+ * Attaches previously detached with [detachObjectGraph] object subgraph.
  * Please note, that once object graph is attached, the stable pointer does not have sense anymore,
  * and shall be discarded.
  */
-fun <T> attachObjectGraph(stable: COpaquePointer?): T =
-        @Suppress("UNCHECKED_CAST")(attachObjectGraphInternal(stable) as T)
+inline fun <reified T> attachObjectGraph(stable: COpaquePointer?): T =
+        @Suppress("UNCHECKED_CAST", "NON_PUBLIC_CALL_FROM_PUBLIC_INLINE")(attachObjectGraphInternal(stable) as T)
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/ObjectTransfer.kt]: `UNCHECKED_CAST` doesn't seem to be required anymore.
`NON_PUBLIC_CALL_FROM_PUBLIC_INLINE ` can be removed, if `attachObjectGraphInternal` is `internal` and marked with `@PublishedApi` annotation.

**Pull request merged**